// @shopify/shop-minis-react/src/hooks/product/useProductListActions.example.tsx
import {useProductListActions, Button} from '@shopify/shop-minis-react'

export default function MyComponent() {
  const {
    addProductList,
    removeProductList,
    renameProductList,
    addProductListItem,
    removeProductListItem,
  } = useProductListActions()

  return (
    <>
      <Button
        onPress={() => addProductList({name: 'My product list'})}
        text="Add product list"
      />
      <Button
        onPress={() => removeProductList({id: 'gid://shopapp/ProductList/123'})}
        text="Remove product list"
      />
      <Button
        onPress={() =>
          renameProductList({
            id: 'gid://shopapp/ProductList/123',
            name: 'My renamed product list',
          })
        }
        text="Rename product list"
      />
      <Button
        onPress={() =>
          addProductListItem({
            shopId: 'gid://shopify/Shop/42',
            productId: 'gid://shopify/Product/123',
            productVariantId: 'gid://shopify/ProductVariant/101',
            listId: 'gid://shopapp/ProductList/99',
            publicListId: 'abc123',
          })
        }
        text="Add product to list"
      />
      <Button
        onPress={() =>
          removeProductListItem({
            shopId: 'gid://shopify/Shop/42',
            productId: 'gid://shopify/Product/123',
            productVariantId: 'gid://shopify/ProductVariant/101',
            listId: 'gid://shopapp/ProductList/99',
            publicListId: 'abc123',
          })
        }
        text="Remove product from list"
      />
    </>
  )
}
