// @shopify/cli-kit/dist/public/node/themes/factories.js
import { Operation } from '@shopify/cli-kit/node/themes/types';
export function buildTheme(themeJson) {
    if (!themeJson)
        return;
    themeJson.processing ?? (themeJson.processing = false);
    themeJson.createdAtRuntime ?? (themeJson.createdAtRuntime = false);
    const { id, name, role, processing, createdAtRuntime } = themeJson;
    return {
        id,
        name,
        processing,
        createdAtRuntime,
        role: role === 'main' ? 'live' : role,
    };
}
export function buildChecksum(asset) {
    if (!asset)
        return;
    const { key, checksum } = asset;
    return { key, checksum };
}
export function buildThemeAsset(asset) {
    if (!asset)
        return;
    const { key, checksum, attachment, value } = asset;
    // Note: for attachments, this is the size of the base64 string, not the real length of the file
    const stats = { size: (value || attachment || '').length, mtime: Date.now() };
    return { key, checksum, attachment, value, stats };
}
export function buildBulkUploadResults(bulkUploadResponse, assets) {
    if (!bulkUploadResponse) {
        return [];
    }
    return bulkUploadResponse.map((bulkUpload, index) => {
        return {
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            key: assets[index].key,
            success: bulkUpload.code === 200,
            errors: bulkUpload.body.errors || {},
            asset: bulkUpload.body.asset,
            operation: Operation.Upload,
        };
    });
}
