// lodash@4.17.21/debounce.js
import { __require as j } from "./isObject.js";
import { __require as A } from "./now.js";
import { __require as D } from "./toNumber.js";
var h, p;
function z() {
  if (p) return h;
  p = 1;
  var W = j(), v = A(), b = D(), C = "Expected a function", L = Math.max, M = Math.min;
  function N(q, r, u) {
    var a, o, l, f, n, t, d = 0, x = !1, c = !1, g = !0;
    if (typeof q != "function")
      throw new TypeError(C);
    r = b(r) || 0, W(u) && (x = !!u.leading, c = "maxWait" in u, l = c ? L(b(u.maxWait) || 0, r) : l, g = "trailing" in u ? !!u.trailing : g);
    function T(e) {
      var i = a, m = o;
      return a = o = void 0, d = e, f = q.apply(m, i), f;
    }
    function R(e) {
      return d = e, n = setTimeout(s, r), x ? T(e) : f;
    }
    function S(e) {
      var i = e - t, m = e - d, k = r - i;
      return c ? M(k, l - m) : k;
    }
    function E(e) {
      var i = e - t, m = e - d;
      return t === void 0 || i >= r || i < 0 || c && m >= l;
    }
    function s() {
      var e = v();
      if (E(e))
        return I(e);
      n = setTimeout(s, S(e));
    }
    function I(e) {
      return n = void 0, g && a ? T(e) : (a = o = void 0, f);
    }
    function y() {
      n !== void 0 && clearTimeout(n), d = 0, a = t = o = n = void 0;
    }
    function O() {
      return n === void 0 ? f : I(v());
    }
    function _() {
      var e = v(), i = E(e);
      if (a = arguments, o = this, t = e, i) {
        if (n === void 0)
          return R(t);
        if (c)
          return clearTimeout(n), n = setTimeout(s, r), T(t);
      }
      return n === void 0 && (n = setTimeout(s, r)), f;
    }
    return _.cancel = y, _.flush = O, _;
  }
  return h = N, h;
}
export {
  z as __require
};
