// @shopify/shop-minis-react/src/hooks/user/useRecentProducts.ts
import {useMemo} from 'react'

import {useShopActions} from '../../internal/useShopActions'
import {useShopActionsPaginatedDataFetching} from '../../internal/useShopActionsPaginatedDataFetching'
import {
  PaginatedDataHookOptionsBase,
  PaginatedDataHookReturnsBase,
  Product,
} from '../../types'

export interface UseRecentProductsParams extends PaginatedDataHookOptionsBase {
  includeSensitive?: boolean
}

export interface UseRecentProductsReturns extends PaginatedDataHookReturnsBase {
  products: Product[] | null
}

/**
 * Hook to fetch recently viewed products.
 * @param params - Options for the hook.
 */
export const useRecentProducts = (
  params?: UseRecentProductsParams
): UseRecentProductsReturns => {
  const {getRecentProducts} = useShopActions()
  const {skip, ...shopActionParams} = params ?? {}

  const {data, ...rest} = useShopActionsPaginatedDataFetching(
    getRecentProducts,
    shopActionParams,
    {
      skip,
      hook: 'useRecentProducts',
    }
  )

  const products = useMemo(() => {
    return data ?? null
  }, [data])

  return {
    ...rest,
    products,
  }
}
