// @shopify/shop-minis-cli/build/commands/codemod/releases/v0-0-93/shoppable-videos/App-after.js
export const shoppableVideosAppTsxAfter = `import {useMemo} from 'react'
import {StatusBar} from 'react-native'
import {
  ShoppableVideos,
  VideoCollectionV2EntryPointItem,
  useMinisParams,
  videoCollectionV2EntrypointFixture,
  MINIS_SANDBOX_SHOP_GID,
  Box,
} from '@shopify/shop-minis-platform-sdk'

export function App() {
  const {entryPointParams, entryPoint: _entryPoint} = useMinisParams()

  // Fixtures provided in the SDK will get you started without entrypoints.
  // Remove them once you have you have your test store setup with your own entrypoints.
  const entryPoint = _entryPoint || videoCollectionV2EntrypointFixture
  const shopGID = entryPointParams?.shopGID || MINIS_SANDBOX_SHOP_GID

  const shoppableVideos = useMemo(() => {
    const collectionItems =
      entryPoint?.type === 'VIDEO_COLLECTION_V2'
        ? entryPoint.collectionItems
        : []
    return mapEntryPointCollectionItemsToShoppableVideos(collectionItems)
  }, [entryPoint])

  return (
    <Box backgroundColor="foregrounds-contrasting-inverted" flex={1}>
      <StatusBar barStyle="light-content" />
      <ShoppableVideos items={shoppableVideos} shopId={shopGID} />
    </Box>
  )
}

function mapEntryPointCollectionItemsToShoppableVideos(
  collectionItems: VideoCollectionV2EntryPointItem[]
) {
  return collectionItems.map(item => ({
    externalId: item.externalId,
    videoUrl: item.video.embeddedUrl,
    fallbackImageUrl: item.fallbackImage.url,
    relatedProductsIds: item.relatedProducts.map(
      relatedProduct => relatedProduct.productId
    ),
  }))
}
`;
