// @shopify/shop-minis-cli/build/commands/codemod/releases/v0-0-93/shoppable-videos/App-before.js
export const shoppableVideosAppTsxBefore = `import {
  ShoppableVideos,
  VideoCollectionV2EntryPointItem,
  useMinisParams,
  videoCollectionV2EntrypointFixture,
  MINIS_SANDBOX_SHOP_GID,
} from '@shopify/shop-minis-platform-sdk'
import {useMemo} from 'react'

export function App() {
  const {entryPointParams, entryPoint: _entryPoint} = useMinisParams()

  // Fixtures provided in the SDK will get you started without entrypoints.
  // Remove them once you have you have your test store setup with your own entrypoints.
  const entryPoint = _entryPoint || videoCollectionV2EntrypointFixture
  const shopGID = entryPointParams?.shopGID || MINIS_SANDBOX_SHOP_GID

  const shoppableVideos = useMemo(() => {
    const collectionItems =
      entryPoint?.type === 'VIDEO_COLLECTION_V2'
        ? entryPoint.collectionItems
        : []
    return mapEntryPointCollectionItemsToShoppableVideos(collectionItems)
  }, [entryPoint])

  return <ShoppableVideos items={shoppableVideos} shopId={shopGID} />
}

function mapEntryPointCollectionItemsToShoppableVideos(
  collectionItems: VideoCollectionV2EntryPointItem[]
) {
  return collectionItems.map(item => ({
    externalId: item.externalId,
    videoUrl: item.video.embeddedUrl,
    fallbackImageUrl: item.fallbackImage.url,
    relatedProductsIds: item.relatedProducts.map(
      relatedProduct => relatedProduct.productId
    ),
  }))
}
`;
