// @shopify/shop-minis-cli/build/constants/extensions.js
export var ExtensionType;
(function (ExtensionType) {
    ExtensionType["BLANK"] = "blank";
    ExtensionType["LINK"] = "link";
    ExtensionType["DEFAULT_CARD"] = "default-card";
    ExtensionType["PRODUCT_OFFER_CARD"] = "product-offer-card";
    ExtensionType["VIDEO_COLLECTION"] = "video-collection";
    ExtensionType["IMAGE_COLLECTION"] = "image-collection";
    ExtensionType["BUNDLE_SELECTOR"] = "bundle-selector";
    ExtensionType["BUNDLE_COLLECTION"] = "bundle-collection";
    ExtensionType["BUNDLE_UPSELL"] = "bundle-upsell";
    ExtensionType["SURVEY_SINGLE_RESPONSE"] = "survey-single-response";
})(ExtensionType || (ExtensionType = {}));
// TODO: import from a single source of truth
export var ExtensionTarget;
(function (ExtensionTarget) {
    ExtensionTarget["shop.product.variants.render-before"] = "shop.product.variants.render-before";
    ExtensionTarget["shop.product.policies.render-after"] = "shop.product.policies.render-after";
    ExtensionTarget["shop.product.block.render"] = "shop.product.block.render";
    ExtensionTarget["shop.store.block.render"] = "shop.store.block.render";
    ExtensionTarget["shop.order-confirmation-page.order-details.render-after"] = "shop.order-confirmation-page.order-details.render-after";
    ExtensionTarget["shop.order-management.order-details.render-after"] = "shop.order-management.order-details.render-after";
    ExtensionTarget["shop.order-management.visit-shop.render-after"] = "shop.order-management.visit-shop.render-after";
})(ExtensionTarget || (ExtensionTarget = {}));
