// @shopify/shop-minis-cli/templates/extensions/components/__template_image-collection/seed.ts
const seedData = {
  title: 'Image Collection Extension',
  items: [
    {
      image: {
        url: 'https://cdn.shopify.com/s/files/1/0621/0463/3599/files/preview_images/197d24d529fd4d5685d88b8e7a0e99b3.thumbnail.0000000000.jpg',
      },
      relatedProducts: [],
      externalId: '1',
    },
    {
      image: {
        url: 'https://cdn.shopify.com/s/files/1/0621/0463/3599/files/preview_images/60d478149138421e94b2a50ba2cac7f3.thumbnail.0000000000.jpg',
      },
      relatedProducts: [],
      externalId: '2',
    },
    {
      image: {
        url: 'https://cdn.shopify.com/s/files/1/0621/0463/3599/files/preview_images/5dfdf5b856cc4e52a92880ab43412db5.thumbnail.0000000000.jpg',
      },
      relatedProducts: [],
      externalId: '3',
    },
    {
      externalId: '4',
      image: {
        url: 'https://cdn.shopify.com/s/files/1/0621/0463/3599/files/preview_images/7a92d6b2112b497fae2bca267ba6dfb1.thumbnail.0000000000.jpg',
      },
      relatedProducts: [],
    },
    {
      externalId: '5',
      image: {
        url: 'https://cdn.shopify.com/s/files/1/0621/0463/3599/files/preview_images/197d24d529fd4d5685d88b8e7a0e99b3.thumbnail.0000000000.jpg',
      },
      relatedProducts: [],
    },
    {
      externalId: '6',
      image: {
        url: 'https://cdn.shopify.com/s/files/1/0621/0463/3599/files/preview_images/197d24d529fd4d5685d88b8e7a0e99b3.thumbnail.0000000000.jpg',
      },
      relatedProducts: [],
    },
    {
      externalId: '7',
      image: {
        url: 'https://cdn.shopify.com/s/files/1/0621/0463/3599/files/preview_images/197d24d529fd4d5685d88b8e7a0e99b3.thumbnail.0000000000.jpg',
      },
      relatedProducts: [],
    },
    {
      externalId: '8',
      image: {
        url: 'https://cdn.shopify.com/s/files/1/0621/0463/3599/files/preview_images/197d24d529fd4d5685d88b8e7a0e99b3.thumbnail.0000000000.jpg',
      },
      relatedProducts: [],
    },
    {
      externalId: '9',
      image: {
        url: 'https://cdn.shopify.com/s/files/1/0621/0463/3599/files/preview_images/197d24d529fd4d5685d88b8e7a0e99b3.thumbnail.0000000000.jpg',
      },
      relatedProducts: [],
    },
  ],
}

export default seedData
