// @shopify/shop-minis-cli/build/commands/codemod/releases/v0-0-93/shoppable-posts/App-after.js
export const shoppablePostsAppTsxAfter = `import {useMemo} from 'react'
import {StatusBar} from 'react-native'
import {
  ImageCollectionV2EntryPointItem,
  ShoppablePosts,
  useMinisParams,
  imageCollectionV2EntrypointFixture,
  MINIS_SANDBOX_SHOP_GID,
  SafeAreaView,
  Box,
} from '@shopify/shop-minis-platform-sdk'

export function App() {
  const {entryPointParams, entryPoint: _entryPoint} = useMinisParams()

  // Fixtures provided in the SDK will get you started without entrypoints.
  // Remove them once you have you have your test store setup with your own entrypoints.
  const entryPoint = _entryPoint || imageCollectionV2EntrypointFixture
  const shopGID = entryPointParams?.shopGID || MINIS_SANDBOX_SHOP_GID

  const shoppablePosts = useMemo(() => {
    const collectionItems =
      entryPoint?.type === 'IMAGE_COLLECTION_V2'
        ? entryPoint.collectionItems
        : []
    return mapEntryPointCollectionItemsToShoppablePosts(collectionItems)
  }, [entryPoint])

  return (
    <Box backgroundColor="backgrounds-regular" flex={1}>
      <StatusBar barStyle="dark-content" />
      <SafeAreaView style={{flex: 1}}>
        <ShoppablePosts items={shoppablePosts} shopId={shopGID} />
      </SafeAreaView>
    </Box>
  )
}

function mapEntryPointCollectionItemsToShoppablePosts(
  collectionItems: ImageCollectionV2EntryPointItem[]
) {
  return collectionItems.map(item => ({
    externalId: item.externalId,
    image: {url: item.image.url},
    description: item.description,
    postedAt: item.displayDate,
    relatedProductsIds: item.relatedProducts.map(
      relatedProduct => relatedProduct.productId
    ),
  }))
}
`;
