// @shopify/shop-minis-cli/.eslintrc.cjs
module.exports = {
  root: true,
  extends: [
    '@react-native-community',
    'plugin:@shopify/typescript',
    'plugin:@shopify/prettier',
  ],
  parserOptions: {
    ecmaVersion: 6,
    project: './tsconfig.json',
  },
  rules: {
    '@babel/no-unused-expressions': 0,
    '@typescript-eslint/no-unused-expressions': [
      'error',
      {
        allowShortCircuit: true,
      },
    ],
    'line-comment-position': 0,
    'no-implicit-coercion': 2,
    'consistent-return': 0,
    'no-console': 0,
    'spaced-comment': 0,
    'no-warning-comments': 'off',
    'no-process-env': 'off',
    'lines-around-comment': 'off',
    '@shopify/typescript/prefer-pascal-case-enums': 'off',
    '@typescript-eslint/naming-convention': 'off',
    '@typescript-eslint/ban-ts-comment': 'off',
    '@typescript-eslint/consistent-indexed-object-style': 'off',
    '@typescript-eslint/member-ordering': 'off',
    'import/no-named-as-default': 'off',
    'import/named': 'off',
    'id-blacklist': [2, 'blacklist', 'whitelist'],
    '@typescript-eslint/camelcase': 'off',
    '@typescript-eslint/ban-ts-ignore': 'off',
    '@typescript-eslint/ban-types': 'off',
    '@typescript-eslint/class-name-casing': 'off',
    '@typescript-eslint/array-type': 'off',
    '@typescript-eslint/no-namespace': 'off',
    '@typescript-eslint/no-unused-vars': [
      'error',
      {
        argsIgnorePattern: '^_',
        varsIgnorePattern: '^_.{1,}',
      },
    ],
    '@typescript-eslint/no-shadow': 'error',
    'no-shadow': 'off',
    'max-len': [
      'error',
      {
        code: 200,
        ignoreUrls: true,
      },
    ],

    // Type files generated by codegen have eslint disabled with a comment
    'eslint-comments/no-unlimited-disable': 'off',
  },
  overrides: [
    {
      files: ['*.ts'],
      rules: {
        '@typescript-eslint/no-non-null-assertion': 0,
        'shopify/typescript/prefer-pascal-case-enums': 0,
      },
    },
    {
      files: ['*.spec.*', '*.test.*'],
      rules: {
        'require-await': 2,
        '@typescript-eslint/no-floating-promises': 2,
      },
    },
  ],
}
