// @shopify/shop-minis-cli/build/commands/submit/utils/request-entry-point-params.js
import { outputInfo } from '@shopify/cli-kit/node/output';
import { keypress } from '@shopify/cli-kit/node/ui';
import { edit } from 'external-editor';
import { getDevelopmentEntryPointParams } from '../../../utils/development-params.js';
export async function requestEntryPointParams() {
    outputInfo('Describe the change being submitted for review. (press enter to launch your preferred editor)');
    await keypress();
    const entryPointParams = edit(getEntryPointParamsTemplate(), {
        postfix: '.yml',
    });
    return cleanUpEntryPointParams(entryPointParams);
}
const EMPTY_ID = 'XXXXXXXXX';
const TEMPLATE_HEADER = '# Params persisted in the file entry-point-params.json are prepopulated here';
/**
 * Removes remaining template strings in case the partners forget to remove them.
 */
function cleanUpEntryPointParams(entryPointParams) {
    return entryPointParams
        .replace(TEMPLATE_HEADER, '')
        .split('\n')
        .filter(line => line.trim().length > 0)
        .filter(line => !line.includes(EMPTY_ID))
        .join('\n');
}
function getEntryPointParamsTemplate() {
    const { external_id, order_id, product_id, product_variant_id, shop_id, shop_domain, } = getDevelopmentEntryPointParams();
    return `${TEMPLATE_HEADER}

- shop_id: ${shop_id || EMPTY_ID}
  product_id: ${product_id || EMPTY_ID}
  product_variant_id: ${product_variant_id || EMPTY_ID}
  external_id: ${external_id || EMPTY_ID}
  order_id: ${order_id || EMPTY_ID}
  shop_domain: ${shop_domain || EMPTY_ID}
- shop_id: ${EMPTY_ID}
  product_id: ${EMPTY_ID}
  product_variant_id: ${EMPTY_ID}
  external_id: ${EMPTY_ID}
  order_id: ${EMPTY_ID}
  shop_domain: ${EMPTY_ID}`;
}
