// @shopify/shop-minis-react/src/hooks/util/useImagePicker.doc.tsx
export const useImagePickerDoc = {
  name: 'useImagePicker',
  category: 'Utility',
  description:
    'Hook for selecting images from camera or gallery in web-based mini apps',
  params: [
    {
      name: 'options',
      type: 'UseImagePickerOptions',
      description: 'Configuration options for the image picker',
      optional: true,
      properties: [
        {
          name: 'cameraFacing',
          type: "'user' | 'environment'",
          description: 'Which camera to use when opening camera',
          optional: true,
          default: "'environment'",
        },
      ],
    },
  ],
  returns: [
    {
      name: 'openCamera',
      type: '() => Promise<File>',
      description: 'Function to open the camera for capturing an image',
    },
    {
      name: 'openGallery',
      type: '() => Promise<File>',
      description: 'Function to open the gallery for selecting an image',
    },
    {
      name: 'ImagePickerInputs',
      type: 'React.FC',
      description:
        'Component that renders hidden file inputs. Must be rendered in your component for the picker to work',
    },
  ],
}
