// @shopify/shop-minis-cli/build/commands/codemod/releases/v0-0-105.js
import { readFile, writeFile } from 'node:fs/promises';
import path from 'node:path';
import { SemVer } from 'semver';
/**
 * Removes ad-hoc eslint config in favour of importing it from the minis runtime
 * Adds graphqlrc config file to minis
 *
 * Diff: https://github.com/Shopify/shop-minis-cli/compare/v0.0.104...v0.0.105
 */
async function applyUpgrade() {
    const eslintConfigPath = path.join(process.cwd(), '.eslintrc.js');
    try {
        const eslintConfig = await readFile(eslintConfigPath, {
            encoding: 'utf-8',
        });
        // only replace the eslint config if it was not changed
        // partners that modified it will need to update manually
        const updatedEslintConfig = eslintConfig.replace(legacyEslintConfig, newEslintConfig);
        await writeFile(eslintConfigPath, updatedEslintConfig, {
            encoding: 'utf-8',
        });
        const graphqlConfigPath = path.join(process.cwd(), '.graphqlrc.js');
        await writeFile(graphqlConfigPath, graphqlConfig, {
            encoding: 'utf-8',
        });
    }
    catch {
        // do nothing
    }
}
const graphqlConfig = "module.exports = require('@shopify/shop-minis-runtime/graphqlrc')\n";
const legacyEslintConfig = `module.exports = {
  root: true,
  extends: ['plugin:@shopify/shop-minis/all'],
  settings: {
    react: {
      version: 'detect',
    },
  },
  parserOptions: {
    ecmaVersion: 6,
    project: './tsconfig.json',
    schema:
      'node_modules/@shopify/shop-minis-platform-sdk/src/api/schema/minis.graphql',
    operations: '**/*.graphql',
  },
}
`;
const newEslintConfig = `const {minisEslintConfig} = require('@shopify/shop-minis-runtime/eslint-config')

module.exports = minisEslintConfig
`;
const exports = {
    releaseNotes: ['Updates eslint config', 'Updates graphqlrc config'],
    applyUpgrade,
    version: new SemVer('0.0.105'),
};
export default exports;
