// @shopify/shop-minis-cli/build/commands/codemod/releases/v0-0-54.js
import fs from 'node:fs';
import path from 'node:path';
import { SemVer } from 'semver';
/**
 * Upgrades shop-minis-runtime version and use ^ for it to autoupgrade on each shop-minis upgrade
 * Remove sdk from dependencies as it's now included in the runtime
 * Upgrade minis eslint-plugin
 * Add resolutions and overrides to package.json to fix errors during npm install and yarn install
 *
 * Diff: https://github.com/Shopify/shop-minis-cli/compare/v0.0.53...v0.0.54
 */
async function applyUpgrade() {
    const packageJsonPath = path.join(process.cwd(), 'package.json');
    const packageJsonString = fs.readFileSync(packageJsonPath, {
        encoding: 'utf-8',
    });
    const packageJson = JSON.parse(packageJsonString) || {};
    // upgrade runtime and use ^
    packageJson.dependencies ||= {};
    packageJson.dependencies['@shopify/shop-minis-runtime'] = '^0.0.7';
    // remove sdk as its now included in runtime
    packageJson.devDependencies ||= {};
    delete packageJson.devDependencies['@shopify/shop-minis-platform-sdk'];
    // upgrade minis eslint-plugin
    packageJson.devDependencies['@shopify/eslint-plugin-shop-minis'] = '^0.0.4';
    // add resolutions and overrides
    packageJson.overrides ||= {};
    packageJson.overrides.react = '18.2.0';
    packageJson.resolutions ||= {};
    packageJson.resolutions['@apollo/federation'] = '0.38.1';
    // persist updated package.json
    fs.writeFileSync(packageJsonPath, `${JSON.stringify(packageJson, null, 2)}\n`, // keep the new line at the end of the file
    {
        encoding: 'utf-8',
    });
    // persist updated .eslintrc.js
    fs.writeFileSync(path.join(process.cwd(), '.eslintrc.js'), newEslintRcContent, {
        encoding: 'utf-8',
    });
}
const releaseNotes = [
    'Upgrade @shopify/shop-minis-runtime and mins eslint-plugin dependencies',
    'Remove @shopify/shop-minis-platform-sdk as it is now included in @shopify/shop-minis-runtime',
    'Add resolutions and overrides to package.json to fix errors when running "npm install" or "yarn install"',
    'Replaced the content of .eslintrc.js as custom rules were moved into the minis preset (If you had custom rules, you will need to add them back manually)',
];
const newEslintRcContent = `module.exports = {
  root: true,
  extends: ['plugin:@shopify/shop-minis/all'],
  settings: {
    react: {
      version: 'detect',
    },
  },
  parserOptions: {
    ecmaVersion: 6,
    project: './tsconfig.json',
    schema:
      'node_modules/@shopify/shop-minis-platform-sdk/src/api/minis.graphql',
    operations: '**/*.graphql',
  },
}
`;
const exports = {
    releaseNotes,
    applyUpgrade,
    version: new SemVer('0.0.54'),
};
export default exports;
