// @shopify/shop-minis-cli/build/commands/codemod/releases/v0-0-114.js
import { readFile, writeFile } from 'node:fs/promises';
import path from 'node:path';
import { SemVer } from 'semver';
import { getPackageLatestVersionCommand } from '../../../utils/package-manager.js';
import { execAsync } from '../../../utils/exec-async-child-process.js';
/**
 * Add shop-minis-extensions and shop-minis-sdk as direct dependencies of the mini
 *
 * Diff: https://github.com/Shopify/shop-minis-cli/compare/v0.0.113...v0.0.114
 */
async function applyUpgrade() {
    const packageJsonPath = path.join(process.cwd(), 'package.json');
    const packageJsonString = await readFile(packageJsonPath, {
        encoding: 'utf-8',
    });
    const packageJson = JSON.parse(packageJsonString) || {};
    packageJson.dependencies = {
        '@shopify/shop-minis-platform-sdk': await fetchLatestVersion('@shopify/shop-minis-platform-sdk', '0.1.96'),
        '@shopify/shop-minis-runtime': packageJson.dependencies['@shopify/shop-minis-runtime'],
        '@shopify/shop-minis-ui-extensions': await fetchLatestVersion('@shopify/shop-minis-ui-extensions', '0.0.5'),
    };
    // persist updated package.json
    await writeFile(packageJsonPath, `${JSON.stringify(packageJson, null, 2)}\n`, // keep the new line at the end of the file
    {
        encoding: 'utf-8',
    });
}
const exports = {
    releaseNotes: [
        '@shopify/shop-minis-platform-sdk and @shopify/shop-minis-ui-extensions are now listed in your project dependencies',
    ],
    applyUpgrade,
    version: new SemVer('0.0.114'),
};
export default exports;
async function fetchLatestVersion(packageName, fallbackVersion) {
    try {
        const getLatestVersionCommand = await getPackageLatestVersionCommand(packageName);
        const { stdout: latestMinisRuntimeVersionStdout } = await execAsync({
            cmd: getLatestVersionCommand,
        });
        return latestMinisRuntimeVersionStdout[0];
    }
    catch {
        return fallbackVersion;
    }
}
