// @shopify/shop-minis-cli/build/commands/codemod/releases/v0-0-84.js
import fs from 'node:fs';
import path from 'node:path';
import { SemVer } from 'semver';
/**
 * Replaces the content of .eslintrc.js with the updated location for minis.graqhl
 *
 * Diff: https://github.com/Shopify/shop-minis-cli/compare/v0.0.83...v0.0.84
 */
async function applyUpgrade() {
    // persist updated .eslintrc.js
    fs.writeFileSync(path.join(process.cwd(), '.eslintrc.js'), newEslintRcContent, {
        encoding: 'utf-8',
    });
}
const releaseNotes = [
    'Replaced the content of .eslintrc.js with the updated location for minis.graqhl',
];
const newEslintRcContent = `module.exports = {
  root: true,
  extends: ['plugin:@shopify/shop-minis/all'],
  settings: {
    react: {
      version: 'detect',
    },
  },
  parserOptions: {
    ecmaVersion: 6,
    project: './tsconfig.json',
    schema:
      'node_modules/@shopify/shop-minis-platform-sdk/src/api/schema/minis.graphql',
    operations: '**/*.graphql',
  },
}
`;
const exports = {
    releaseNotes,
    applyUpgrade,
    version: new SemVer('0.0.84'),
};
export default exports;
