// motion@12.17.3/dist/es/motion-dom/dist/es/animation/generators/spring/find.js
import { springDefaults as n } from "./defaults.js";
import { warning as u } from "../../../../../../motion-utils/dist/es/errors.js";
import { secondsToMilliseconds as g, millisecondsToSeconds as N } from "../../../../../../motion-utils/dist/es/time-conversion.js";
import { clamp as h } from "../../../../../../motion-utils/dist/es/clamp.js";
const D = 1e-3;
function v({ duration: t = n.duration, bounce: p = n.bounce, velocity: a = n.velocity, mass: e = n.mass }) {
  let i, f;
  u(t <= g(n.maxDuration), "Spring duration must be 10 seconds or less");
  let o = 1 - p;
  o = h(n.minDamping, n.maxDamping, o), t = h(n.minDuration, n.maxDuration, N(t)), o < 1 ? (i = (s) => {
    const r = s * o, c = r * t, l = r - a, m = b(s, o), x = Math.exp(-c);
    return D - l / m * x;
  }, f = (s) => {
    const c = s * o * t, l = c * a + a, m = Math.pow(o, 2) * Math.pow(s, 2) * t, x = Math.exp(-c), S = b(Math.pow(s, 2), o);
    return (-i(s) + D > 0 ? -1 : 1) * ((l - m) * x) / S;
  }) : (i = (s) => {
    const r = Math.exp(-s * t), c = (s - a) * t + 1;
    return -1e-3 + r * c;
  }, f = (s) => {
    const r = Math.exp(-s * t), c = (a - s) * (t * t);
    return r * c;
  });
  const w = 5 / t, M = A(i, f, w);
  if (t = g(t), isNaN(M))
    return {
      stiffness: n.stiffness,
      damping: n.damping,
      duration: t
    };
  {
    const s = Math.pow(M, 2) * e;
    return {
      stiffness: s,
      damping: o * 2 * Math.sqrt(e * s),
      duration: t
    };
  }
}
const T = 12;
function A(t, p, a) {
  let e = a;
  for (let i = 1; i < T; i++)
    e = e - t(e) / p(e);
  return e;
}
function b(t, p) {
  return t * Math.sqrt(1 - p * p);
}
export {
  b as calcAngularFreq,
  v as findSpring
};
