// @shopify/shop-minis-cli/templates/extensions/components/__template_video-collection/seed.ts
const seedData = {
  title: 'Video Collection Extension',
  items: [
    {
      video: {
        url: 'https://cdn.shopify.com/videos/c/o/v/197d24d529fd4d5685d88b8e7a0e99b3.mp4',
      },
      fallbackImage: {
        url: 'https://cdn.shopify.com/s/files/1/0621/0463/3599/products/Main_9129b69a-0c7b-4f66-b6cf-c4222f18028a.jpg?v=1671634025',
      },
      relatedProducts: [],
      externalId: '',
    },
    {
      video: {
        url: 'https://cdn.shopify.com/videos/c/o/v/5dfdf5b856cc4e52a92880ab43412db5.mp4',
      },
      fallbackImage: {
        url: 'https://cdn.shopify.com/s/files/1/0621/0463/3599/products/Main_0a4e9096-021a-4c1e-8750-24b233166a12.jpg?v=1671633756',
      },
      relatedProducts: [],
      externalId: '',
    },
    {
      video: {
        url: 'https://cdn.shopify.com/videos/c/o/v/60d478149138421e94b2a50ba2cac7f3.mp4',
      },
      fallbackImage: {
        url: 'https://cdn.shopify.com/s/files/1/0621/0463/3599/products/Main_c8ff0b5d-c712-429a-be00-b29bd55cbc9d.jpg?v=1671634817',
      },
      relatedProducts: [],
      externalId: '',
    },
  ],
}

export default seedData
