// @shopify/shop-minis-react/dist/components/atoms/accordion.js
import { jsx as e, jsxs as a } from "react/jsx-runtime";
import { cn as r } from "../../lib/utils.js";
import i from "../../node_modules/.pnpm/lucide-react@0.513.0_react@19.1.0/node_modules/lucide-react/dist/esm/icons/chevron-down.js";
import { Root as s, Item as d, Header as c, Trigger as l, Content as m } from "../../node_modules/.pnpm/@radix-ui_react-accordion@1.2.11_@types_react-dom@19.1.6_@types_react@19.1.6__@types_react@19_krqvbhitcyb3vl2jzw3fp35uhu/node_modules/@radix-ui/react-accordion/dist/index.js";
function b({
  ...t
}) {
  return /* @__PURE__ */ e(s, { "data-slot": "accordion", ...t });
}
function x({
  className: t,
  ...o
}) {
  return /* @__PURE__ */ e(
    d,
    {
      "data-slot": "accordion-item",
      className: r("border-b last:border-b-0", t),
      ...o
    }
  );
}
function v({
  className: t,
  children: o,
  ...n
}) {
  return /* @__PURE__ */ e(c, { className: "flex", children: /* @__PURE__ */ a(
    l,
    {
      "data-slot": "accordion-trigger",
      className: r(
        "focus-visible:border-ring focus-visible:ring-ring/50 flex flex-1 items-start justify-between gap-4 rounded-md py-4 text-left text-sm font-medium transition-all outline-none hover:underline focus-visible:ring-[3px] disabled:pointer-events-none disabled:opacity-50 [&[data-state=open]>svg]:rotate-180",
        t
      ),
      ...n,
      children: [
        o,
        /* @__PURE__ */ e(i, { className: "text-muted-foreground pointer-events-none size-4 shrink-0 translate-y-0.5 transition-transform duration-200" })
      ]
    }
  ) });
}
function h({
  className: t,
  children: o,
  ...n
}) {
  return /* @__PURE__ */ e(
    m,
    {
      "data-slot": "accordion-content",
      className: "data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down overflow-hidden text-sm",
      ...n,
      children: /* @__PURE__ */ e("div", { className: r("pt-0 pb-4", t), children: o })
    }
  );
}
export {
  b as Accordion,
  h as AccordionContent,
  x as AccordionItem,
  v as AccordionTrigger
};
