// @shopify/shop-minis-react/dist/components/atoms/alert-dialog.js
import { jsx as e, jsxs as r } from "react/jsx-runtime";
import { cn as o } from "../../lib/utils.js";
import { buttonVariants as l } from "./button.js";
import { Root as n, Portal as i, Overlay as s, Trigger as d, Content as c, Title as g, Description as u, Action as f, Cancel as m } from "../../node_modules/.pnpm/@radix-ui_react-alert-dialog@1.1.14_@types_react-dom@19.1.6_@types_react@19.1.6__@types_react_5smkpul3xqqoqjowwwfk226yey/node_modules/@radix-ui/react-alert-dialog/dist/index.js";
function b({
  ...t
}) {
  return /* @__PURE__ */ e(n, { "data-slot": "alert-dialog", ...t });
}
function y({
  ...t
}) {
  return /* @__PURE__ */ e(d, { "data-slot": "alert-dialog-trigger", ...t });
}
function p({
  ...t
}) {
  return /* @__PURE__ */ e(i, { "data-slot": "alert-dialog-portal", ...t });
}
function x({
  className: t,
  ...a
}) {
  return /* @__PURE__ */ e(
    s,
    {
      "data-slot": "alert-dialog-overlay",
      className: o(
        "data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 fixed inset-0 z-50 bg-black/50",
        t
      ),
      ...a
    }
  );
}
function T({
  className: t,
  ...a
}) {
  return /* @__PURE__ */ r(p, { children: [
    /* @__PURE__ */ e(x, {}),
    /* @__PURE__ */ e(
      c,
      {
        "data-slot": "alert-dialog-content",
        className: o(
          "bg-background data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 fixed top-[50%] left-[50%] z-50 grid w-full max-w-[calc(100%-2rem)] translate-x-[-50%] translate-y-[-50%] gap-4 rounded-lg border p-6 shadow-lg duration-200 sm:max-w-lg",
          t
        ),
        ...a
      }
    )
  ] });
}
function w({ className: t, ...a }) {
  return /* @__PURE__ */ e(
    "div",
    {
      "data-slot": "alert-dialog-header",
      className: o("flex flex-col gap-2 text-center sm:text-left", t),
      ...a
    }
  );
}
function C({ className: t, ...a }) {
  return /* @__PURE__ */ e(
    "div",
    {
      "data-slot": "alert-dialog-footer",
      className: o(
        "flex flex-col-reverse gap-2 sm:flex-row sm:justify-end",
        t
      ),
      ...a
    }
  );
}
function z({
  className: t,
  ...a
}) {
  return /* @__PURE__ */ e(
    g,
    {
      "data-slot": "alert-dialog-title",
      className: o("text-lg font-semibold", t),
      ...a
    }
  );
}
function h({
  className: t,
  ...a
}) {
  return /* @__PURE__ */ e(
    u,
    {
      "data-slot": "alert-dialog-description",
      className: o("text-muted-foreground text-sm", t),
      ...a
    }
  );
}
function j({
  className: t,
  ...a
}) {
  return /* @__PURE__ */ e(
    f,
    {
      className: o(l(), t),
      ...a
    }
  );
}
function O({
  className: t,
  ...a
}) {
  return /* @__PURE__ */ e(
    m,
    {
      className: o(l({ variant: "outlined" }), t),
      ...a
    }
  );
}
export {
  b as AlertDialog,
  j as AlertDialogAction,
  O as AlertDialogCancel,
  T as AlertDialogContent,
  h as AlertDialogDescription,
  C as AlertDialogFooter,
  w as AlertDialogHeader,
  x as AlertDialogOverlay,
  p as AlertDialogPortal,
  z as AlertDialogTitle,
  y as AlertDialogTrigger
};
