// @shopify/shop-minis-react/dist/components/atoms/avatar.js
import { jsx as e } from "react/jsx-runtime";
import { cn as r } from "../../lib/utils.js";
import { Root as l, Image as o, Fallback as s } from "../../node_modules/.pnpm/@radix-ui_react-avatar@1.1.10_@types_react-dom@19.1.6_@types_react@19.1.6__@types_react@19.1._hkz57sehyui4ndfh3rsqwxftli/node_modules/@radix-ui/react-avatar/dist/index.js";
function u({
  className: a,
  ...t
}) {
  return /* @__PURE__ */ e(
    l,
    {
      "data-slot": "avatar",
      className: r(
        "relative flex size-8 shrink-0 overflow-hidden rounded-full",
        a
      ),
      ...t
    }
  );
}
function m({
  className: a,
  ...t
}) {
  return /* @__PURE__ */ e(
    o,
    {
      "data-slot": "avatar-image",
      className: r("aspect-square size-full", a),
      ...t
    }
  );
}
function c({
  className: a,
  ...t
}) {
  return /* @__PURE__ */ e(
    s,
    {
      "data-slot": "avatar-fallback",
      className: r(
        "bg-muted flex size-full items-center justify-center rounded-full",
        a
      ),
      ...t
    }
  );
}
export {
  u as Avatar,
  c as AvatarFallback,
  m as AvatarImage
};
