// @shopify/shop-minis-react/dist/components/atoms/dialog.js
import { jsx as a, jsxs as n } from "react/jsx-runtime";
import { cn as o } from "../../lib/utils.js";
import r from "../../node_modules/.pnpm/lucide-react@0.513.0_react@19.1.0/node_modules/lucide-react/dist/esm/icons/x.js";
import { Root as d, Close as l, Content as g, Description as c, Overlay as u, Portal as f, Title as m, Trigger as p } from "../../node_modules/.pnpm/@radix-ui_react-dialog@1.1.14_@types_react-dom@19.1.6_@types_react@19.1.6__@types_react@19.1._sjczexpvrqz6fttoobpwnns2oa/node_modules/@radix-ui/react-dialog/dist/index.js";
function N({ ...t }) {
  return /* @__PURE__ */ a(d, { "data-slot": "dialog", ...t });
}
function z({
  ...t
}) {
  return /* @__PURE__ */ a(p, { "data-slot": "dialog-trigger", ...t });
}
function x({
  ...t
}) {
  return /* @__PURE__ */ a(f, { "data-slot": "dialog-portal", ...t });
}
function w({
  ...t
}) {
  return /* @__PURE__ */ a(l, { "data-slot": "dialog-close", ...t });
}
function v({
  className: t,
  ...e
}) {
  return /* @__PURE__ */ a(
    u,
    {
      "data-slot": "dialog-overlay",
      className: o(
        "data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 fixed inset-0 z-50 bg-black/50",
        t
      ),
      ...e
    }
  );
}
function C({
  className: t,
  children: e,
  showCloseButton: i = !0,
  ...s
}) {
  return /* @__PURE__ */ n(x, { "data-slot": "dialog-portal", children: [
    /* @__PURE__ */ a(v, {}),
    /* @__PURE__ */ n(
      g,
      {
        "data-slot": "dialog-content",
        className: o(
          "bg-background data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 fixed top-[50%] left-[50%] z-50 grid w-full max-w-[calc(100%-2rem)] translate-x-[-50%] translate-y-[-50%] gap-4 rounded-lg border p-6 shadow-lg duration-200 sm:max-w-lg",
          t
        ),
        ...s,
        children: [
          e,
          i && /* @__PURE__ */ n(
            l,
            {
              "data-slot": "dialog-close",
              className: "ring-offset-background focus:ring-ring data-[state=open]:bg-accent data-[state=open]:text-muted-foreground absolute top-4 right-4 rounded-xs opacity-70 transition-opacity hover:opacity-100 focus:ring-2 focus:ring-offset-2 focus:outline-hidden disabled:pointer-events-none [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4",
              children: [
                /* @__PURE__ */ a(r, {}),
                /* @__PURE__ */ a("span", { className: "sr-only", children: "Close" })
              ]
            }
          )
        ]
      }
    )
  ] });
}
function k({ className: t, ...e }) {
  return /* @__PURE__ */ a(
    "div",
    {
      "data-slot": "dialog-header",
      className: o("flex flex-col gap-2 text-center sm:text-left", t),
      ...e
    }
  );
}
function T({ className: t, ...e }) {
  return /* @__PURE__ */ a(
    "div",
    {
      "data-slot": "dialog-footer",
      className: o(
        "flex flex-col-reverse gap-2 sm:flex-row sm:justify-end",
        t
      ),
      ...e
    }
  );
}
function j({
  className: t,
  ...e
}) {
  return /* @__PURE__ */ a(
    m,
    {
      "data-slot": "dialog-title",
      className: o("text-lg leading-none font-semibold", t),
      ...e
    }
  );
}
function _({
  className: t,
  ...e
}) {
  return /* @__PURE__ */ a(
    c,
    {
      "data-slot": "dialog-description",
      className: o("text-muted-foreground text-sm", t),
      ...e
    }
  );
}
export {
  N as Dialog,
  w as DialogClose,
  C as DialogContent,
  _ as DialogDescription,
  T as DialogFooter,
  k as DialogHeader,
  v as DialogOverlay,
  x as DialogPortal,
  j as DialogTitle,
  z as DialogTrigger
};
