// @shopify/shop-minis-react/dist/components/atoms/radio-group.js
import { jsx as r } from "react/jsx-runtime";
import { cn as o } from "../../lib/utils.js";
import t from "../../node_modules/.pnpm/lucide-react@0.513.0_react@19.1.0/node_modules/lucide-react/dist/esm/icons/circle.js";
import { Root as e, Item as s, Indicator as d } from "../../node_modules/.pnpm/@radix-ui_react-radio-group@1.3.7_@types_react-dom@19.1.6_@types_react@19.1.6__@types_react@1_b6vd3n6tkje342bk7jtbsrqyfe/node_modules/@radix-ui/react-radio-group/dist/index.js";
function m({
  className: i,
  ...a
}) {
  return /* @__PURE__ */ r(
    e,
    {
      "data-slot": "radio-group",
      className: o("grid gap-3", i),
      ...a
    }
  );
}
function p({
  className: i,
  ...a
}) {
  return /* @__PURE__ */ r(
    s,
    {
      "data-slot": "radio-group-item",
      className: o(
        "border-input text-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 aspect-square size-4 shrink-0 rounded-full border shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50",
        i
      ),
      ...a,
      children: /* @__PURE__ */ r(
        d,
        {
          "data-slot": "radio-group-indicator",
          className: "relative flex items-center justify-center",
          children: /* @__PURE__ */ r(t, { className: "fill-primary absolute top-1/2 left-1/2 size-2 -translate-x-1/2 -translate-y-1/2" })
        }
      )
    }
  );
}
export {
  m as RadioGroup,
  p as RadioGroupItem
};
