// @shopify/shop-minis-react/dist/components/atoms/scroll-area.js
import { jsxs as t, jsx as l } from "react/jsx-runtime";
import { cn as a } from "../../lib/utils.js";
import { Root as s, Viewport as c, Corner as n, ScrollAreaScrollbar as i, ScrollAreaThumb as u } from "../../node_modules/.pnpm/@radix-ui_react-scroll-area@1.2.9_@types_react-dom@19.1.6_@types_react@19.1.6__@types_react@1_na6tcikkwmgt5sgxl6ezf7zaf4/node_modules/@radix-ui/react-scroll-area/dist/index.js";
function p({
  className: o,
  children: r,
  ...e
}) {
  return /* @__PURE__ */ t(
    s,
    {
      "data-slot": "scroll-area",
      className: a("relative", o),
      ...e,
      children: [
        /* @__PURE__ */ l(
          c,
          {
            "data-slot": "scroll-area-viewport",
            className: "focus-visible:ring-ring/50 size-full rounded-[inherit] transition-[color,box-shadow] outline-none focus-visible:ring-[3px] focus-visible:outline-1",
            children: r
          }
        ),
        /* @__PURE__ */ l(d, {}),
        /* @__PURE__ */ l(n, {})
      ]
    }
  );
}
function d({
  className: o,
  orientation: r = "vertical",
  ...e
}) {
  return /* @__PURE__ */ l(
    i,
    {
      "data-slot": "scroll-area-scrollbar",
      orientation: r,
      className: a(
        "flex touch-none p-px transition-colors select-none",
        r === "vertical" && "h-full w-2.5 border-l border-l-transparent",
        r === "horizontal" && "h-2.5 flex-col border-t border-t-transparent",
        o
      ),
      ...e,
      children: /* @__PURE__ */ l(
        u,
        {
          "data-slot": "scroll-area-thumb",
          className: "bg-border relative flex-1 rounded-full"
        }
      )
    }
  );
}
export {
  p as ScrollArea,
  d as ScrollBar
};
