// graphql-typescript-definitions/build/cjs/filesystem/default-graphql-filesystem.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var chokidar = require('chokidar');
var graphqlConfigUtilities = require('graphql-config-utilities');
var graphqlFilesystem = require('./graphql-filesystem.js');

class DefaultGraphQLFilesystem extends graphqlFilesystem.AbstractGraphQLFilesystem {
  constructor(...args) {
    super(...args);
    this.watchers = [];
  }
  async watch(config) {
    this.watchers.push(...this.setupDocumentWatchers(config).concat(this.setupSchemaWatcher(config)));
    await Promise.all(this.watchers.map(watcher => new Promise(resolve => watcher.on('ready', () => resolve()))));
  }
  dispose() {
    this.watchers.forEach(watcher => {
      watcher.close();
    });
    this.watchers.length = 0;
  }
  getGraphQLProjectIncludedFilePaths(projectConfig) {
    return graphqlConfigUtilities.getGraphQLProjectIncludedFilePaths(projectConfig);
  }
  setupDocumentWatchers(config) {
    return graphqlConfigUtilities.getGraphQLProjects(config).filter(projectConfig => {
      const [includes] = graphqlConfigUtilities.getIncludesExcludesFromConfig(projectConfig);
      return includes.length > 0;
    }).map(projectConfig => {
      const [includes, excludes] = graphqlConfigUtilities.getIncludesExcludesFromConfig(projectConfig);
      return chokidar.watch(includes.map(include => graphqlConfigUtilities.resolvePathRelativeToConfig(projectConfig, include)), {
        ignored: excludes.map(exclude => graphqlConfigUtilities.resolvePathRelativeToConfig(projectConfig, exclude)),
        ignoreInitial: true
      }).on('add', filePath => this.emit('change:document', filePath, projectConfig)).on('change', filePath => this.emit('change:document', filePath, projectConfig)).on('unlink', filePath => this.emit('delete:document', filePath, projectConfig));
    });
  }
  setupSchemaWatcher(config) {
    return chokidar.watch(graphqlConfigUtilities.getGraphQLSchemaPaths(config), {
      ignoreInitial: true
    }).on('change', schemaPath => this.emit('change:schema', schemaPath));
  }
}

exports.DefaultGraphQLFilesystem = DefaultGraphQLFilesystem;
