// motion@12.17.3/dist/es/motion-dom/dist/es/animation/generators/keyframes.js
import { interpolate as c } from "../../utils/interpolate.js";
import { defaultOffset as p } from "../keyframes/offsets/default.js";
import { convertOffsetToTimes as u } from "../keyframes/offsets/time.js";
import { easeInOut as l } from "../../../../../motion-utils/dist/es/easing/ease.js";
import { isEasingArray as g } from "../../../../../motion-utils/dist/es/easing/utils/is-easing-array.js";
import { easingDefinitionToFunction as m } from "../../../../../motion-utils/dist/es/easing/utils/map.js";
function T(n, t) {
  return n.map(() => t || l).splice(0, n.length - 1);
}
function y({ duration: n = 300, keyframes: t, times: o, ease: r = "easeInOut" }) {
  const e = g(r) ? r.map(m) : m(r), i = {
    done: !1,
    value: t[0]
  }, s = u(
    // Only use the provided offsets if they're the correct length
    // TODO Maybe we should warn here if there's a length mismatch
    o && o.length === t.length ? o : p(t),
    n
  ), a = c(s, t, {
    ease: Array.isArray(e) ? e : T(t, e)
  });
  return {
    calculatedDuration: n,
    next: (f) => (i.value = a(f), i.done = f >= n, i)
  };
}
export {
  T as defaultEasing,
  y as keyframes
};
